L1Trigger/L1TTrackerDTC/src/Settings.cc

```diff
@@ -0,0 +1,418 @@
+    // derived Router parameter
+    numDTCs_ = numRegions_ * numDTCsPerRegion_;
+    numModules_ = numRegions_ * numDTCsPerRegion_ * numModulesPerDTC_;
+    numModulesPerRoutingBlock_ = numModulesPerDTC_ / numRoutingBlocks_;
+
+    const int numFramesPerBX = freqDTC_ / freqLHC_;
+    maxFramesChannelInput_ = numFramesPerBX * tmpFE_ - numFramesInfra_;
+    maxFramesChannelOutput_ = numFramesPerBX * tmpTFP_ - numFramesInfra_;
+
+    numMergedRows_ = pow(2, widthRow_ - widthRowLUT_);
+    maxCot_ = sinh(maxEta_);
+
+    rangeQoverPt_ = speedOfLight_ * bField_ / minPt_ / 1000.;
+
+    widthLayer_ = ceil(log2(numLayers_));
+
+    baseRegion_ = 2. * M_PI / numRegions_;
+
+    if (dataFormat_ == "TMTT")
+      tmtt_ = new SettingsTMTT(iConfig, this);
+    else if (dataFormat_ == "Hybrid")
+      hybrid_ = new SettingsHybrid(iConfig, this);
+    else
+      throw cms::Exception("L1TrackerDTC::Settings::Settings unknown data format requested (" + dataFormat_ + ").");
+
+    maxQoverPt_ = (rangeQoverPt_ - baseQoverPt_) / 2.;
+
+    const int baseDiffM = widthRowLUT_ - widthRow_;
+
+    baseM_ = basePhi_ * pow(2., baseDiffM);
+    baseC_ = basePhi_;
+
+    widthC_ = widthPhi_;
+
+    const double x1 = pow(2, widthRow_) * baseRow_ * maxPitch_ / 2.;
+    const double x0 = x1 - pow(2, widthRowLUT_) * baseRow_ * maxPitch_;
+    const double maxM = atan2(x1, innerRadius_) - atan2(x0, innerRadius_);
+
+    widthM_ = ceil(log2(maxM / baseM_));
+
+    // event setup
+    cablingMap_ = vector<DetId>(numModules_);
+    trackerGeometry_ = nullptr;
+    trackerTopology_ = nullptr;
+  }
+
+  Settings::~Settings() {
+    if (dataFormat_ == "TMTT")
+      delete tmtt_;
+    else if (dataFormat_ == "Hybrid")
+      delete hybrid_;
+  }
+
+  // read in detector parameter
+  void Settings::beginRun(const Run& iRun, const EventSetup& iSetup) {
+    // read in MagneticField
+    ESHandle<MagneticField> magneticFieldHandle;
+    iSetup.get<IdealMagneticFieldRecord>().get(magneticFieldHandle);
+    const MagneticField* magneticField = magneticFieldHandle.product();
+
+    // read in TrackerGeometry
+    ESHandle<TrackerGeometry> trackerGeometryHandle;
+    iSetup.get<TrackerDigiGeometryRecord>().get(trackerGeometryHandle);
+    trackerGeometry_ = trackerGeometryHandle.product();
+
+    // read in TrackerTopology
+    ESHandle<TrackerTopology> trackerTopologyHandle;
+    iSetup.get<TrackerTopologyRcd>().get(trackerTopologyHandle);
+    trackerTopology_ = trackerTopologyHandle.product();
+
+    // read in cabling map
+    ESHandle<TrackerDetToDTCELinkCablingMap> cablingMapHandle;
+    iSetup.get<TrackerDetToDTCELinkCablingMapRcd>().get(cablingMapHandle);
+    const TrackerDetToDTCELinkCablingMap* cablingMap = cablingMapHandle.product();
+
+    // check magnetic field configuration
+    const double bField = magneticField->inTesla(GlobalPoint(0., 0., 0.)).z();
+    if (fabs(bField - bField_) > bFieldError_) {
+      cms::Exception exception("Configuration",
```

Comment:
why does this need to be checked?

Reply:
well it would be good to recognize when the firmware thinks the bfield would be significant smaller or bigger as that what cms thinks.

Reply:
I don't quite understand the reasoning. Is this something the firmware actually checks in real life? In the simulation, the magnetic field is exactly what is in the database record.

Reply:
the firmware has no access to the database record, the bfield is hard coded there.
If the DB changes I would like recognize that. We could remove the check, but then we have to hope that we don't forget to update the f/w when DB changes.